File: bin/dww/bindings/niche/darkwow.zig
// DarkWow Wallet — Zig FFI binding
//   const darkwow = @import("darkwow");
//   var w = try darkwow.Wallet.open("keys.toml", "wallet-1", "testnet");
//   defer w.close();
//   const n = w.scanBlock(block_json);
//   std.debug.print("scanned {d} outputs, balance={d}\n", .{ n, w.balance() });

const std = @import("std");

const WalletHandle = opaque {};

extern "c" fn dwow_wallet_version() [*c]const u8;
extern "c" fn dwow_wallet_open(keys_path: [*c]const u8, section: [*c]const u8, network: [*c]const u8) ?*WalletHandle;
extern "c" fn dwow_wallet_free(handle: ?*WalletHandle) void;
extern "c" fn dwow_wallet_scan_block_json(handle: ?*WalletHandle, block_json: [*c]const u8) c_int;
extern "c" fn dwow_wallet_cap_count(handle: ?*WalletHandle) c_int;
extern "c" fn dwow_wallet_balance(handle: ?*WalletHandle) u64;
extern "c" fn dwow_wallet_chain_height(handle: ?*WalletHandle) u64;

pub const Wallet = struct {
    handle: *WalletHandle,

    pub fn open(keys_path: [:0]const u8, section: [:0]const u8, network: [:0]const u8) !Wallet {
        const h = dwow_wallet_open(keys_path.ptr, section.ptr, network.ptr) orelse
            return error.OpenFailed;
        return Wallet{ .handle = h };
    }

    pub fn close(self: *Wallet) void {
        dwow_wallet_free(self.handle);
    }

    pub fn scanBlock(self: *Wallet, json: [:0]const u8) c_int {
        return dwow_wallet_scan_block_json(self.handle, json.ptr);
    }

    pub fn capCount(self: *Wallet) c_int { return dwow_wallet_cap_count(self.handle); }
    pub fn balance(self: *Wallet) u64 { return dwow_wallet_balance(self.handle); }
    pub fn chainHeight(self: *Wallet) u64 { return dwow_wallet_chain_height(self.handle); }

    pub fn version() [:0]const u8 {
        return std.mem.span(dwow_wallet_version());
    }
};

File: bin/dww/bindings/niche/darkwow.hs
-- DarkWow Wallet — Haskell FFI binding
--   import DarkWow
--   main = do
--     w <- openWallet "keys.toml" "wallet-1" "testnet"
--     n <- scanBlock w blockJson
--     putStrLn $ "scanned " ++ show n ++ " outputs, balance=" ++ show (balance w)

{-# LANGUAGE ForeignFunctionInterface #-}

module DarkWow
    ( WalletHandle
    , openWallet, closeWallet
    , scanBlock, capCount, balance, chainHeight
    , version
    ) where

import Foreign.C.String (CString, newCString, peekCString, withCString)
import Foreign.C.Types (CInt(..), CULLong(..))
import Foreign.Ptr (Ptr, nullPtr)
import Control.Exception (bracket)

data WalletHandle
type WalletPtr = Ptr WalletHandle

foreign import ccall "dwow_wallet_version"   c_version :: IO CString
foreign import ccall "dwow_wallet_open"       c_open :: CString -> CString -> CString -> IO WalletPtr
foreign import ccall "dwow_wallet_free"       c_free :: WalletPtr -> IO ()
foreign import ccall "dwow_wallet_scan_block_json" c_scan :: WalletPtr -> CString -> IO CInt
foreign import ccall "dwow_wallet_cap_count"   c_capCount :: WalletPtr -> IO CInt
foreign import ccall "dwow_wallet_balance"    c_balance :: WalletPtr -> IO CULLong
foreign import ccall "dwow_wallet_chain_height" c_chainHeight :: WalletPtr -> IO CULLong

openWallet :: String -> String -> String -> IO WalletPtr
openWallet keysPath section network =
    withCString keysPath $ \kp ->
    withCString section  $ \sec ->
    withCString network  $ \net -> do
        h <- c_open kp sec net
        if h == nullPtr then error "dwow_wallet_open failed" else return h

closeWallet :: WalletPtr -> IO ()
closeWallet = c_free

scanBlock :: WalletPtr -> String -> IO Int
scanBlock h json = withCString json $ \j -> fromIntegral <$> c_scan h j

capCount :: WalletPtr -> IO Int
capCount h = fromIntegral <$> c_capCount h

balance :: WalletPtr -> IO Integer
balance h = fromIntegral <$> c_balance h

chainHeight :: WalletPtr -> IO Integer
chainHeight h = fromIntegral <$> c_chainHeight h

version :: IO String
version = peekCString =<< c_version

-- Recommended usage: bracket (openWallet ...) closeWallet $ \w -> do ...
